### PR TITLE
add `as_sequence()` method on dict views

### PR DESCRIPTION
**Pull request by PrettyWood**
Hi and thank you for this amazing library! I'm trying to help @samuelcolvin on [https://github.com/samuelcolvin/pydantic-core](pydantic-core). Following https://github.com/PyO3/pyo3/pull/2358, it would be great to have at list a `as_sequence()` method on dict views to be able to easily get length, convert them to list, iterate...
I didn't want to add any other methods following @davidhewitt 's comment. Let's keep it simple for now :)

 - [x] an entry in CHANGELOG.md
 - [x] docs to all new functions and / or detail in the guide
 - [x] tests for all new or changed functions

### CODE REVIEW

**Review by samuelcolvin — APPROVED** [review]
LGTM :+1:

**Comment by PrettyWood** [comment]
I don't know what's wrong on windows and if it's even related to this PR 🤔

File: src/types/dict.rs [hunk 1]
```diff
@@ -46,6 +57,13 @@ pyobject_native_type_core!(
     #checkfunction=ffi::PyDictValues_Check
 );
 
+#[cfg(not(PyPy))]
+impl PyDictValues {
+    pub fn as_sequence(&self) -> &PySequence {
+        unsafe { PySequence::try_from_unchecked(self) }
+    }
+}
+
```
Comment (mejrs) [inline src/types/dict.rs:66]: This should have a comment similar to the existing `as_sequence` methods (same for PyDictItems)

**Review by mejrs — COMMENTED** [review]
Thanks!

> I don't know what's wrong on windows and if it's even related to this PR 🤔

Looks like `The filename or extension is too long. (os error 206)`...ugh. I don't know what's up with that.

**Comment by PrettyWood** [comment]
dankjewel @mejrs for the quick review! 🙏

**Comment by davidhewitt** [comment]
Hey, thanks for this. 

I'm definitely willing to add some methods to these views to improve usability, although I'm not convinced `.as_sequence()` is strictly correct. This appears to break type-safety, as I don't think these collections are sequences:

```python
>>> import collections
>>> isinstance({}.keys(), collections.abc.Sequence)
False
```

You can already make use of the fact these views implement `Deref<Target=PyAny>` to use `.len()`, `.iter()` etc.

If we add methods to these types I would guess the main motivation would either be:
- performance if there's a C API we can make use of
- remove need for `PyResult<_>` on `.len()` etc.

Is it possible that someone could subtype `dict_keys` (e.g.) and create a `__len__` which raises an exception? I think so, in which case we have to make a judgement call whether a bespoke `fn len() -> usize` on these types would panic in that case. So far in PyO3 we've made the decision to return results in all such cases so as to minimise library-level panics.

**Comment by PrettyWood** [comment]
Ah you're absolutely right @davidhewitt! Thank you!
Since order is preserved, I thought dict views would be sequences but since they are not we can close this PR.
----------------------
At first I also added a `len()` method indeed to avoid having a `Result` type but I understand the judgement call.

Thanks again for this library!

**Pull request closed without merging**